src/brod_sock.erl

```diff
@@ -203,6 +206,38 @@ maybe_upgrade_to_ssl(Sock, _Mod = ssl, SslOpts = [_|_], Timeout) ->
 maybe_upgrade_to_ssl(Sock, _Mod, _SslOpts, _Timeout) ->
   {ok, Sock}.
 
+maybe_sasl_auth(_Sock, _Mod, _ClientId, _Timeout, _SaslOpts = undefined) -> ok;
+maybe_sasl_auth(Sock, Mod, ClientId, Timeout,
+                _SaslOpts = {_Method = "PLAIN", SaslUser, SaslPassword}) ->
+  ok = setopts(Sock, Mod, [{active, false}]),
+  HandshakeRequest = #kpro_SaslHandshakeRequest{mechanism="PLAIN"},
+  HandshakeRequestBin = kpro:encode_request(ClientId, 0, HandshakeRequest),
+  ok = Mod:send(Sock, HandshakeRequestBin),
+  {ok, <<Len:32>>} = Mod:recv(Sock, 4, Timeout),
+  {ok, HandshakeResponseBin} = Mod:recv(Sock, Len, Timeout),
+  {[ #kpro_Response{ responseMessage = #kpro_SaslHandshakeResponse{
+                                          errorCode = ErrorCode }}],
+    <<>>} = kpro:decode_response(<<Len:32, HandshakeResponseBin/binary>>),
+  case ErrorCode of
+    no_error ->
+      SaslMessage = list_to_binary([0, unicode:characters_to_binary(SaslUser),
```

Comment:
no need to construct binary.
the second arg of ssl:send/2 allows `iodata() `
http://erlang.org/doc/man/ssl.html#send-2